sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala

```diff
@@ -125,12 +129,37 @@ case class InMemoryTableScanExec(
     val schema = relation.partitionStatistics.schema
     val schemaIndex = schema.zipWithIndex
     val relOutput: AttributeSeq = relation.output
-    val buffers = relation.cachedColumnBuffers
+    val partitionFilter = newPredicate(
+      partitionFilters.reduceOption(And).getOrElse(Literal(true)),
+      schema)
+
+    val buffers = if (inMemoryPartitionPruningEnabled && !relation.batchStats.value.isEmpty) {
+      val validPartitions = relation.batchStats.value.asScala
+        .filter(batchStat => partitionFilter(batchStat._2))
+        .map(_._1)
+        .distinct
+      if (validPartitions.isEmpty) {
+        sparkContext.emptyRDD[CachedBatch]
+      } else {
+        PartitionPruningRDD.create[CachedBatch](relation.cachedColumnBuffers,
+          index => {
+            if (validPartitions.contains(index)) {
+              true
+            } else {
+              logInfo(s"Skipping partition $index because all cached batches will be pruned")
```

Comment:
Current executor-side pruning logging is done at INFO. I have no strong opinion either way, but this can get noisy with many partitions getting pruned.